pyvrp/cpp/educate/LocalSearch.cpp

```diff
@@ -192,21 +193,20 @@ void LocalSearch::maybeInsert(Node *U,
 {
     assert(!U->route && V->route);
 
-    auto const &uClient = data.client(U->client);
-
-    int const current = data.dist(V->client, n(V)->client);
-    int const proposed = data.dist(V->client, U->client)
-                         + data.dist(U->client, n(V)->client) - uClient.prize;
+    distance_type const deltaDist = data.dist(V->client, U->client)
+                                    + data.dist(U->client, n(V)->client)
+                                    - data.dist(V->client, n(V)->client);
 
-    int deltaCost = proposed - current;
+    auto const &uClient = data.client(U->client);
+    cost_type deltaCost = static_cast<cost_type>(deltaDist) - uClient.prize;
 
     deltaCost += costEvaluator.loadPenalty(V->route->load() + uClient.demand,
                                            data.vehicleCapacity());
     deltaCost
         -= costEvaluator.loadPenalty(V->route->load(), data.vehicleCapacity());
 
-    if (deltaCost >= V->route->timeWarp())  // adding U will likely not lower
-        return;                             // time warp, so we can stop here.
+    if (deltaCost >= 0)  // adding U will likely not lower time warp,
```

Comment:
As currently written, it is probably not correct. I suspect it should have been something like
```cpp
if (deltaCost >= costEvaluator.twPenalty(V->route->timeWarp()))
```
but that condition is rather weak. We could strengthen it to
```cpp
if (
    (V->route->isFeasible() && deltaCost >= 0) 
    || (deltaCost >= costEvaluator.twPenalty(V->route->timeWarp())
)
    return;
```
?

Reply:
Hmm indeed the penalty is missing there as well. Shall we fix this in a seperate PR?

I think the two options you propose are equivalent. Adding the first operand doesn't change anything, since if the route is feasible, the timewarp is 0 and so is the penalty, so the second operand will effectively be `deltaCost >= 0`. Or am I missing something?

In general the current version generalizes the `(V->route->isFeasible() && deltaCost >= 0)` check. I implemented similar checks for `Exchange` in https://github.com/PyVRP/PyVRP/pull/50 and here is a [diff](https://github.com/PyVRP/PyVRP/compare/separate-distance-duration...cost-evaluator-bounding) which shows how we could consider bounding more generally (also valid if we would include duration in the objective).

Reply:
Yeah, sorry, I missed that. You are right. I'll update this to
```
if (deltaCost >= costEvaluator.twPenalty(V->route->timeWarp()))
```

Reply:
We must update this particular location in this PR already because deltaCost cannot be compared to a Duration type; it needs to be Cost.